crates/typst/src/text/font/color.rs

```diff
@@ -266,3 +272,356 @@ fn make_svg_unsized(svg: &mut String) {
         svg.replace_range(range, "");
     }
 }
+
+struct ColrBuilder<'a>(&'a mut String);
+
+impl ColrBuilder<'_> {
+    fn finish(&mut self) {
+        if !self.0.is_empty() {
+            self.0.pop(); // remove trailing space
+        }
+    }
+}
+
+impl ttf_parser::OutlineBuilder for ColrBuilder<'_> {
+    fn move_to(&mut self, x: f32, y: f32) {
+        use std::fmt::Write;
+        write!(self.0, "M {x} {y} ").unwrap()
+    }
+
+    fn line_to(&mut self, x: f32, y: f32) {
+        use std::fmt::Write;
+        write!(self.0, "L {x} {y} ").unwrap()
+    }
+
+    fn quad_to(&mut self, x1: f32, y1: f32, x: f32, y: f32) {
+        use std::fmt::Write;
+        write!(self.0, "Q {x1} {y1} {x} {y} ").unwrap()
+    }
+
+    fn curve_to(&mut self, x1: f32, y1: f32, x2: f32, y2: f32, x: f32, y: f32) {
+        use std::fmt::Write;
+        write!(self.0, "C {x1} {y1} {x2} {y2} {x} {y} ").unwrap()
+    }
+
+    fn close(&mut self) {
+        self.0.push_str("Z ")
+    }
+}
+
+trait XmlWriterExt {
+    fn write_color_attribute(&mut self, name: &str, ts: ttf_parser::RgbaColor);
+    fn write_transform_attribute(&mut self, name: &str, ts: ttf_parser::Transform);
+    fn write_spread_method_attribute(&mut self, method: ttf_parser::colr::GradientExtend);
+}
+
+impl XmlWriterExt for xmlwriter::XmlWriter {
```

Comment:
Why are these not just on `GlyphPainter`?

Reply:
Mainly because I copy-pasted it from the `resvg` implementation, but I will change it.